--- ./Makefile
+++ ./Makefile
@@ -12,11 +12,11 @@
 # LUA_ROOT, LUA_LDIR, and LUA_CDIR in luaconf.h.
 INSTALL_TOP= /usr/local
 INSTALL_BIN= $(INSTALL_TOP)/bin
-INSTALL_INC= $(INSTALL_TOP)/include
-INSTALL_LIB= $(INSTALL_TOP)/lib
-INSTALL_MAN= $(INSTALL_TOP)/man/man1
-INSTALL_LMOD= $(INSTALL_TOP)/share/lua/$V
-INSTALL_CMOD= $(INSTALL_TOP)/lib/lua/$V
+INSTALL_INC= $(INSTALL_TOP)/usr/include
+INSTALL_LIB= $(INSTALL_TOP)/lib64
+INSTALL_MAN= $(INSTALL_TOP)/usr/share/man/man1
+INSTALL_LMOD= $(INSTALL_TOP)/usr/share/lua/$V
+INSTALL_CMOD= $(INSTALL_TOP)/lib64
 
 # How to install. If your install program does not support "-p", then
 # you may have to run ranlib on the installed liblua.a.
--- ./src/luaconf.h
+++ ./src/luaconf.h
@@ -241,9 +241,9 @@
 
 #else			/* }{ */
 
-#define LUA_ROOT	"/usr/local/"
-#define LUA_LDIR	LUA_ROOT "share/lua/" LUA_VDIR "/"
-#define LUA_CDIR	LUA_ROOT "lib/lua/" LUA_VDIR "/"
+#define LUA_ROOT	"/"
+#define LUA_LDIR	"/usr/share/lua/" LUA_VDIR "/"
+#define LUA_CDIR	"/lib64/"
 
 #if !defined(LUA_PATH_DEFAULT)
 #define LUA_PATH_DEFAULT  \
