--- a/passwordsafe/database_manager.py.orig
+++ b/passwordsafe/database_manager.py
@@ -3,7 +3,7 @@
 
 import hashlib
 import logging
-from datetime import datetime
+from datetime import datetime, timezone
 from gettext import gettext as _
 from uuid import UUID
 
@@ -208,11 +208,11 @@
 
     def set_element_atime(self, element):
         # TODO Move to safe_element method.
-        element.atime = datetime.now()
+        element.atime = datetime.now(timezone.utc)
 
     def set_element_mtime(self, element):
         # TODO Move to safe_element method.
-        element.mtime = datetime.now()
+        element.mtime = datetime.now(timezone.utc)
 
     # Move an group
     def move_group(self, group: Group, dest_group: Group) -> None:
--- a/passwordsafe.in.orig
+++ b/passwordsafe.in
@@ -5,6 +5,8 @@
 import locale
 import gettext
 
+os.environ['GTK_THEME'] = 'Adwaita:dark'
+
 import gi
 gi.require_version('Gtk', '3.0')
 gi.require_version("Handy", "1")
--- a/data/meson.build.orig
+++ b/data/meson.build
@@ -35,7 +35,6 @@
 endif
 desktop_conf.set('ICON', application_id)
 desktop = i18n.merge_file(
-  'desktop',
   input: configure_file(
     input: files('org.gnome.PasswordSafe.desktop.in.in'),
     output: 'org.gnome.PasswordSafe.desktop.in',
@@ -65,7 +64,6 @@
 endif
 appdata_conf.set('APPID', application_id)
 appdata = i18n.merge_file(
-  'appdata',
   input: configure_file(
     input: files('org.gnome.PasswordSafe.appdata.xml.in.in'),
     output: 'org.gnome.PasswordSafe.appdata.xml.in',
