--- a/Makefile.orig
+++ b/Makefile
@@ -1,5 +1,5 @@
 CC=gcc
-CFLAGS=-O3 -Wall -Wextra -Wno-unknown-pragmas -Wcast-qual
+CFLAGS=-O3 -march=x86-64-v2 -DNOZOPFLI
 LDFLAGS=
 # CFLAGS=-O3 -Wall -Wextra -Wno-unknown-pragmas -Wcast-qual -g -fsanitize=thread
 # LDFLAGS=-g -fsanitize=thread
@@ -11,11 +11,11 @@
 
 # use gcc and gmake on Solaris
 
-pigz: pigz.o yarn.o try.o $(ZOP)
-	$(CC) $(LDFLAGS) -o pigz pigz.o yarn.o try.o $(ZOP) $(LIBS)
+pigz: pigz.o yarn.o try.o
+	$(CC) $(LDFLAGS) -o pigz pigz.o yarn.o try.o $(LIBS)
 	ln -f pigz unpigz
 
-pigz.o: pigz.c yarn.h try.h $(ZOPFLI)deflate.h $(ZOPFLI)util.h
+pigz.o: pigz.c yarn.h try.h
 
 yarn.o: yarn.c yarn.h
 
@@ -109,3 +109,5 @@
 
 clean:
 	@rm -f *.o pigz unpigz pigzj pigzn pigzt pigz.c.gz pigz.c.zz pigz.c.zip
+
+install:
